api/envoy/extensions/filters/http/custom_response/v3/custom_response.proto

```diff
@@ -0,0 +1,112 @@
+// policies in a matcher tree. The configuration can be defined at the filter,
+// virtual host or route level. The most specific configuration will apply.
+message CustomResponse {
+  // [#next-free-field: 2]
+
+  // Matcher to match against the original response.
+  // Example:
+  //
+  // .. validated-code-block:: yaml
+  //   :type-name: xds.type.matcher.v3.Matcher
+  //
+  //   matcher_list:
+  //     matchers:
+  //       # Apply a locally stored custom response to any 4xx response.
+  //     - predicate:
+  //         single_predicate:
+  //           input:
+  //             name: 4xx_response
+  //             typed_config:
+  //               "@type": type.googleapis.com/envoy.type.matcher.v3.HttpResponseStatusCodeClassMatchInput
+  //           value_match:
+  //             exact: "4xx"
+  //       on_match:
+  //         action:
+  //           name: action
+  //           typed_config:
+  //             "@type": type.googleapis.com/envoy.config.core.v3.TypedExtensionConfig
+  //             name: local_response
+  //             typed_config:
+  //               "@type": type.googleapis.com/envoy.extensions.filters.http.custom_response.v3.LocalResponsePolicy
+  //               status_code: 499
+  //               body:
+  //                 inline_string: "not allowed"
+  //               body_format:
+  //                 json_format:
+  //                   status: "%RESPONSE_CODE%"
+  //                   message: "%LOCAL_REPLY_BODY%"
+  //               response_headers_to_add:
+  //               - header:
+  //                   key: "foo"
+  //                   value: "x-bar"
+  //       # Redirect to different upstream if the status code is one of 502, 503 or 504.
+  //     - predicate:
+  //         or_matcher:
+  //           predicate:
+  //           - single_predicate:
+  //               input:
+  //                 name: "502_response"
+  //                 typed_config:
+  //                   "@type": type.googleapis.com/envoy.type.matcher.v3.HttpResponseStatusCodeMatchInput
+  //               value_match:
+  //                 exact: "502"
+  //           - single_predicate:
+  //               input:
+  //                 name: "503_response"
+  //                 typed_config:
+  //                   "@type": type.googleapis.com/envoy.type.matcher.v3.HttpResponseStatusCodeMatchInput
+  //               value_match:
+  //                 exact: "503"
+  //           - single_predicate:
+  //               input:
+  //                 name: "504_response"
+  //                 typed_config:
+  //                   "@type": type.googleapis.com/envoy.type.matcher.v3.HttpResponseStatusCodeMatchInput
+  //               value_match:
+  //                 exact: "504"
+  //       on_match:
+  //         action:
+  //           name: action
+  //           typed_config:
+  //             "@type": type.googleapis.com/envoy.config.core.v3.TypedExtensionConfig
```

Comment:
I do not think you need this double wrapping of policy config into the TypedExtensionConfig. You can just include the RedirectPolicy directly. This will simplify your factory code for policies.

Reply:
As I understand it, ActionBase takes a proto Type as a template parameter. The action for the custom response matcher can be one of several policies, and the policy definitions need to be wrapped in a single proto definition which will be that template parameter.

Reply:
On second thoughts, I don't see Action::typeUrl being used or enforced anywhere in the matcher infrastructure. Let me take a stab at implementing your suggestion, as it will make the interface and implementation a whole lot cleaner.

Reply:
Thanks so much for your suggestion. I was quite irked by my initial wrapping of TypedExtensionConfigs, but clearly not enough to push through for a better solution.

PTAL